### PR TITLE
Implement all methods of java.util.Comparator

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
I fixed type inference on 2.12. Unfortunately, that meant a lot of type annotations.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ComparatorTest.scala [hunk 1]
```diff
@@ -40,4 +45,190 @@ class ComparatorTest {
...
+    val cmp = ju.Comparator.comparingDouble((_: String).length.toDouble / 2)
+    assertEquals(0, cmp.compare("a", "b"))
+    assertTrue(cmp.compare("", "a") < 0)
+    assertTrue(cmp.compare("ab", "") > 0)
+
+    assertThrowsNPEIfCompliant(ju.Comparator.comparingDouble(null))
+  }
+
+  @Test def thenComparingComparator(): Unit = {
+    val base = ju.Comparator.comparingInt((_: (Int, Int))._1)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ComparatorTest.scala:151]: This is not super readable, IMO. Consider the following style when you have a tuple argument:
```suggestion
    val base = ju.Comparator.comparingInt((x: (Int, Int)) => x._1)
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Looks good overall. I have two small comments.

File: javalib/src/main/scala/java/util/Comparator.scala [hunk 2]
```diff
@@ -12,12 +12,153 @@
...
+  def reverseOrder[T <: Comparable[T]](): Comparator[T] =
+    naturalOrder[T]().reversed()
+
+  /* Should be T <: Comparable[_ >: T] but scalac fails with
+   * > illegal cyclic reference involving type U
+   */
+  @inline
+  def naturalOrder[T <: Comparable[T]](): Comparator[T] = new Comparator[T] with Serializable {
+    def compare(o1: T, o2: T): Int = o1.compareTo(o2)
+  }
```
[the remark this answers is not in the document]
Comment (gzm0) [reply javalib/src/main/scala/java/util/Comparator.scala]: Done. I have also added a commit that replaces an internal version of this we had for collections. LMK if you think it should be squashed.

File: javalib/src/main/scala/java/util/NaturalComparator.scala [hunk 3]
```diff
@@ -12,27 +12,35 @@
...
-    o1.asInstanceOf[Comparable[Any]].compareTo(o2)
+private[util] object NaturalComparator {
+  /** Returns a natural comparator even for types that are not statically
+   *  known to be <: Comparable.
+   */
+  @inline
+  def forAnyType[A](): Comparator[A] = {
+    // Any type argument that is comparable will do.
+    Comparator.naturalOrder[String]().asInstanceOf[Comparator[A]]
+  }
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/NaturalComparator.scala:36]: I don't like that we're using `String` here. Consider adding the following in `Comparator` to make things less ugly:
```scala
  @inline
  private[util] def naturalOrderForAnyType[A](): Comparator[A] =
    ReusableNaturalOrder.asInstanceOf[Comparator[A]]
```
and call that from this `forAnyType` method. This way, we don't have to go through a specific non-`Any` type, and the viability of the casts stays closer to the original definition.

In fact, even `def naturalOrder()` itself can delegate to `naturalOrderForAnyType`, in that case.

**Review by sjrd — CHANGES_REQUESTED** [review]

File: javalib/src/main/scala/java/util/NaturalComparator.scala (hunk 3 above)
Comment (gzm0) [reply javalib/src/main/scala/java/util/NaturalComparator.scala]: Yes, fair enough. I have opted to just use `NaturalComparator` as it is. That also gets rid of the strange contract that `forAnyType` must return the same object all the time (for `select` / `unselect`).

**Review by sjrd — APPROVED** [review]

**Pull request merged**